File: Blind150/6.LinkedList/GO/reorder_list.go
// https://leetcode.com/problems/remove-nth-node-from-end-of-list/description/
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reorderList(head *ListNode) {

	s, f := head, head

	for f != nil && f.Next != nil {
		s = s.Next
		f = f.Next.Next

	}

	// find the head of the 2nd List
	snd := s.Next

	s.Next = nil

	// reverse the 2nd List
	prev, tmp := (*ListNode)(nil), (*ListNode)(nil)

	for snd != nil {

		tmp = snd.Next

		snd.Next = prev

		prev = snd

		snd = tmp

	}

	ft, sd := head, prev

	tmp1, tmp2 := (*ListNode)(nil), (*ListNode)(nil)

	for sd != nil {

		tmp1, tmp2 = ft.Next, sd.Next

		ft.Next = sd
		sd.Next = tmp1

		ft, sd = tmp1, tmp2

	}

}

File: Blind150/6.LinkedList/Python/reorder_list.py
# https://leetcode.com/problems/remove-nth-node-from-end-of-list/description/
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:

        s,f = head, head.next

        while f and f.next:

            s = s.next
            f = f.next.next


        snd = s.next

        s.next = None

        # reverse the 2nd half

        prev = None

        while snd:

            tmp = snd.next
            snd.next = prev
            prev = snd
            snd = tmp


        # merge both the lists
        # snd = prev ; because , after the previous loop is done snd will be
        # none and prev will the head of the snd half of the List

        ft, sd = head, prev

        while sd:

            tmp1, tmp2 = ft.next, sd.next

            ft.next = sd
            sd.next = tmp1
            ft,sd = tmp1, tmp2
